File: retroauth-core/src/test/java/com/andretietz/retroauth/RetroauthTest.java
package com.andretietz.retroauth;

import com.andretietz.retroauth.testimpl.TestInterface;
import com.andretietz.retroauth.testimpl.TestProvider;
import com.andretietz.retroauth.testimpl.TestResponse;
import com.andretietz.retroauth.testimpl.TestTokenStorage;

import org.junit.After;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mockito;
import org.mockito.runners.MockitoJUnitRunner;

import java.io.IOException;

import okhttp3.mockwebserver.MockResponse;
import okhttp3.mockwebserver.MockWebServer;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava.HttpException;
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory;
import retrofit2.converter.moshi.MoshiConverterFactory;
import rx.observers.TestSubscriber;
import rx.schedulers.Schedulers;


/**
 * Created by andre on 11.06.2016.
 */
@RunWith(MockitoJUnitRunner.class)
public class RetroauthTest {

    private MockWebServer server;
    private TestInterface service;
    private TestTokenStorage tokenStorage;

    @Before
    public void prepare() throws IOException {
        server = new MockWebServer();
        server.start();
        tokenStorage = new TestTokenStorage();
        MethodCache.DefaultMethodCache<String> methodCache = new MethodCache.DefaultMethodCache<>();
        AuthenticationHandler<String, String, String> authHandler =
                new AuthenticationHandler<>(
                        methodCache,
                        Mockito.mock(OwnerManager.class),
                        tokenStorage,
                        new TestProvider()
                );

        Retrofit retrofit = new Retroauth.Builder<>(authHandler)
                .baseUrl(server.url("/"))
                .enableLocking(true)
                .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
                .addConverterFactory(MoshiConverterFactory.create())
                .build();

        service = retrofit.create(TestInterface.class);
    }

    @After
    public void shutdown() throws IOException {
        server.shutdown();
    }

    @Test
    public void simpleHappyCase() throws IOException {
        // test very simple case
        server.enqueue(new MockResponse().setBody(TestInterface.TEST_BODY));
        TestResponse response = service.authenticatedMethod().toBlocking().single();
        Assert.assertEquals(1, response.value);

        server.enqueue(new MockResponse().setBody(TestInterface.TEST_BODY));
        response = service.unauthenticatedMethod().toBlocking().single();
        Assert.assertEquals(1, response.value);

    }

    @Test
    public void simpleErrorCase() throws IOException {
        TestSubscriber<TestResponse> subscriber = TestSubscriber.create();
        // test simple error case
        server.enqueue(new MockResponse().setResponseCode(400));
        service.authenticatedMethod().subscribe(subscriber);
        subscriber.awaitTerminalEvent();
        subscriber.assertNoValues();
        subscriber.assertError(HttpException.class);

        subscriber = TestSubscriber.create();
        server.enqueue(new MockResponse().setResponseCode(400));
        service.unauthenticatedMethod().subscribe(subscriber);
        subscriber.awaitTerminalEvent();
        subscriber.assertNoValues();
        subscriber.assertError(HttpException.class);
    }

    @Test
    public void userCanceledAuthentication() throws IOException, InterruptedException {
        tokenStorage.setTestBehaviour(new TestTokenStorage.TestBehaviour() {
            @Override
            public String getToken(String owner, String tokenType) throws AuthenticationCanceledException {
                throw new AuthenticationCanceledException("userCanceledAuthentication");
            }
        });

        TestSubscriber<TestResponse> subscriber = TestSubscriber.create();

        service.authenticatedMethod().subscribe(subscriber);
        subscriber.awaitTerminalEvent();
        subscriber.assertNoValues();
        subscriber.assertError(AuthenticationCanceledException.class);

        subscriber = TestSubscriber.create();
        server.enqueue(new MockResponse().setBody(TestInterface.TEST_BODY));
        service.unauthenticatedMethod().subscribe(subscriber);
        subscriber.awaitTerminalEvent();
        subscriber.assertNoErrors();
        subscriber.assertValueCount(1);

        tokenStorage.setTestBehaviour(null);
    }

    /**
     * This test should fail, if you're not using {@link Retroauth.Builder#enableLocking(boolean)}
     * with the value <code>true</code>
     */
    @Test
    public void blockingErrorCaseTest() {
        int requestCount = 100;
        TestSubscriber<TestResponse>[] subscribers = new TestSubscriber[requestCount];
        // create a lot of requests
        for (int i = 0; i < requestCount; i++) {
            subscribers[i] = TestSubscriber.create();
            service.authenticatedMethod()
                    .subscribeOn(Schedulers.newThread())
                    .subscribe(subscribers[i]);
        }
        // create possible failing responses
        for (int i = 0; i < requestCount; i++) {
            server.enqueue(new MockResponse().setResponseCode(401));
        }

        // catch responses
        for (int i = 0; i < requestCount; i++) {
            subscribers[i].awaitTerminalEvent();
            subscribers[i].assertNoValues();
            subscribers[i].assertError(Exception.class);
        }
        // ensure that only one request has been executed
        Assert.assertEquals(1, server.getRequestCount());

    }
}

File: retroauth/src/test/java/com/andretietz/retroauth/RetroauthTest.kt
package com.andretietz.retroauth

import com.andretietz.retroauth.testimpl.TestInterface
import com.andretietz.retroauth.testimpl.TestProvider
import com.andretietz.retroauth.testimpl.TestResponse
import com.andretietz.retroauth.testimpl.TestTokenStorage
import com.andretietz.retroauth.testimpl.TestTokenTypeFactory
import com.nhaarman.mockito_kotlin.eq
import com.nhaarman.mockito_kotlin.spy
import com.nhaarman.mockito_kotlin.times
import com.nhaarman.mockito_kotlin.verify
import com.nhaarman.mockito_kotlin.whenever
import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers.anyString
import org.mockito.Mock
import org.mockito.junit.MockitoJUnitRunner
import retrofit2.HttpException
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.moshi.MoshiConverterFactory
import java.io.IOException


@RunWith(MockitoJUnitRunner::class)
class RetroauthTest {

    private lateinit var server: MockWebServer
    private lateinit var service: TestInterface
    private var provider: TokenProvider<String> = spy(TestProvider())
    private var tokenStorage = spy(TestTokenStorage())

    companion object {
        val expectedResult = TestResponse(1)
    }

    @Mock
    private lateinit var ownerManager: OwnerManager<String, String>

    @Before
    @Throws(IOException::class)
    fun prepare() {
        server = MockWebServer()
        server.start()
        val methodCache = MethodCache.DefaultMethodCache<String>()
        val authHandler = AuthenticationHandler(
                methodCache,
                ownerManager,
                tokenStorage,
                provider,
                TestTokenTypeFactory()
        )

        val retrofit = Retroauth.Builder(authHandler)
                .baseUrl(server.url("/"))
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .addConverterFactory(MoshiConverterFactory.create())
                .build()

        service = retrofit.create(TestInterface::class.java)

        whenever(ownerManager.getOwner(anyString())).thenReturn("owner")
    }

    @After
    fun shutdown() {
        server.shutdown()
    }

    @Test
    fun simpleHappyCase() {
        // test very simple case
        server.enqueue(MockResponse().setBody(TestInterface.TEST_BODY))

        service.authenticatedMethod()
                .test()
                .assertResult(expectedResult)

        server.enqueue(MockResponse().setBody(TestInterface.TEST_BODY))

        service.unauthenticatedMethod()
                .test()
                .assertResult(expectedResult)
    }

    @Test
    fun simpleErrorCase() {
        // test simple error case
        server.enqueue(MockResponse().setResponseCode(400))
        service.authenticatedMethod()
                .test()
                .assertFailure(HttpException::class.java)


        server.enqueue(MockResponse().setResponseCode(400))
        service.unauthenticatedMethod()
                .test()
                .assertFailure(HttpException::class.java)
    }

    @Test
    fun userCanceledAuthentication() {

        whenever(ownerManager.getOwner(anyString())).thenThrow(AuthenticationCanceledException::class.java)

        service.authenticatedMethod()
                .test()
                .assertFailure(AuthenticationCanceledException::class.java)

        server.enqueue(MockResponse().setBody(TestInterface.TEST_BODY))
        service.unauthenticatedMethod()
                .test()
                .assertResult(expectedResult)

    }

    @Test
    fun testIfTokenRefreshWorks() {

        server.enqueue(MockResponse().setResponseCode(401))

        server.enqueue(
                MockResponse()
                        .setResponseCode(200)
                        .setBody(TestInterface.TEST_BODY)
        )

        

        whenever(provider.refreshToken(eq("token")))
                .thenAnswer {
                    Thread.sleep(250)
                    return@thenAnswer "new token"
                }
        service.authenticatedMethod()
                .test()
                .assertResult(expectedResult)

        verify(tokenStorage, times(1)).storeToken(anyString(), anyString(), eq("new token"))


    }
//    @Test
//    fun concurrencyTestForTokenRefreshes() {
//
//        server.enqueue(MockResponse().setResponseCode(401))
//
//        server.enqueue(
//                MockResponse()
//                        .setResponseCode(200)
//                        .setBody(TestInterface.TEST_BODY)
//        )
//
//        whenever(provider.refreshToken(eq("token")))
//                .thenAnswer {
//                    Thread.sleep(250)
//                    return@thenAnswer "new token"
//                }
//        service.authenticatedMethod()
//                .test()
//                .assertResult(expectedResult)
//
//        verify(tokenStorage, times(1)).storeToken(anyString(), anyString(), eq("new token"))
//
//
//    }
}
